### ISSUE TITLE
Salt request timed out. The master is not responding. While running orchestrate only

### ISSUE THREAD

**Issue by Oloremo** [saltstack/salt#50621]
### Description of Issue/Question
I have few formulas which run just fine using the state.apply on 31 minions but fails to run in the orchestrate runner with 31 minions, yet works fine with 5 minions via orchestrate. 

So I assume the number of minions is affecting the result.

### Setup
**cat states/orch/hadoop.sls**
```
---
include:
  - .sync

install.java:
  salt.state:
    - tgt: '*'
    - sls:
      - formula.java

install.zookeeper:
  salt.state:
    - tgt: 'roles:zookeeper'
    - tgt_type: 'grain'
    - sls:
      - formula.zookeeper
    - require:
      - salt: install.java
```
**cat states/orch/sync.sls**
```
salt.mine.update:
  salt.function:
    - tgt: "*"
    - name: mine.update
    - order: 1

salt.sync.all:
  salt.function:
    - tgt: "*"
    - name: saltutil.sync_all
    - order: 1
```

**cat states/formula/java/install.sls**
```
{% from tpldir +"/map.jinja" import java %}
{% set common_user = pillar.get('common_user') %}
{% set archive_urls = pillar.get('archive_urls') %}

{% set release_dir = java.dirs.releases_root + "/" + java.version %}
{% do java.dirs.update({'release_dir': release_dir}) %}

{% for dir_name, dir_path in java.dirs.items() %}
formula.java.dir.{{ dir_name }}:
  file.directory:
    - name: {{ dir_path }}
    - user: {{ common_user.name }}
    - group: {{ common_user.group }}
    - mode: 775
    - makedirs: True
{% endfor %}

formula.java.extract_java:
  archive.extracted:
    - name: {{ java.dirs.release_dir }}
    - user: {{ common_user.name }}
    - group: {{ common_user.group }}
    - source: {{ archive_urls['java']['path'] }}/{{ archive_urls['java']['file'] }}
    - source_hash: {{ archive_urls['java']['path'] }}/{{ archive_urls['java']['file'] }}.md5
    - options: "--strip-components=1"
    - enforce_toplevel: False
    - trim_output: True
    - if_missing: {{ java.dirs.release_dir }}/bin/java
    - require:
        - file: formula.java.dir.release_dir

formula.java.create_symlink:
  file.symlink:
    - name: {{ java.app_current }}
    - target: {{ java.dirs.release_dir }}
    - require:
      - archive: formula.java.extract_java

formula.java.profile:
  file.managed:
    - name: "{{ java.dirs.home }}/.profile"
    - source: salt://{{ tpldir }}/templates/profile.j2
    - user: {{ common_user.name }}
    - group: {{ common_user.group }}
    - mode: 644
    - template: jinja
    - context:
        java: {{ java }}
```

`salt '*' state.apply formula.java` work just fine yet `salt-run state.orchestrate orch.hadoop` fails with the following error:

```
  Name: saltutil.sync_all - Function: salt.function - Result: Changed Started: - 11:57:21.709636 Duration: 1228.042 ms
----------
          ID: install.java
    Function: salt.state
      Result: False
     Comment: An exception occurred in this state: Traceback (most recent call last):
                File "/usr/lib/python2.7/site-packages/salt/state.py", line 1913, in call
                  **cdata['kwargs'])
                File "/usr/lib/python2.7/site-packages/salt/loader.py", line 1898, in wrapper
                  return f(*args, **kwargs)
                File "/usr/lib/python2.7/site-packages/salt/states/saltmod.py", line 333, in state
                  cmd_ret = __salt__['saltutil.cmd'](tgt, fun, **cmd_kw)
                File "/usr/lib/python2.7/site-packages/salt/modules/saltutil.py", line 1399, in cmd
                  client, tgt, fun, arg, timeout, tgt_type, ret, kwarg, **kwargs)
                File "/usr/lib/python2.7/site-packages/salt/modules/saltutil.py", line 1363, in _exec
                  for ret_comp in _cmd(**cmd_kwargs):
                File "/usr/lib/python2.7/site-packages/salt/client/__init__.py", line 615, in cmd_batch
                  batch = salt.cli.batch.Batch(opts, eauth=eauth, quiet=True)
                File "/usr/lib/python2.7/site-packages/salt/cli/batch.py", line 39, in __init__
                  self.minions, self.ping_gen, self.down_minions = self.__gather_minions()
                File "/usr/lib/python2.7/site-packages/salt/cli/batch.py", line 66, in __gather_minions
                  for ret in ping_gen:
                File "/usr/lib/python2.7/site-packages/salt/client/__init__.py", line 910, in cmd_iter
                  **kwargs)
                File "/usr/lib/python2.7/site-packages/salt/client/__init__.py", line 358, in run_job
                  raise SaltClientError(general_exception)
              SaltClientError: Salt request timed out. The master is not responding. You may need to run your command with `--async` in order to bypass the congested event bus. With `--async`, the CLI tool will print the job id (jid) and exit immediately without listening for responses. You can then use `salt-run jobs.lookup_jid` to look up the results of the job in the job cache later.
     Started: 11:57:22.938030
    Duration: 20685.77 ms
     Changes:
----------
          ID: install.zookeeper
    Function: salt.state
      Result: False
     Comment: One or more requisite failed: orch.hadoop.install.java
     Started: 11:57:43.625152
    Duration: 0.023 ms
     Changes:

Summary for delme-salt-rm.aws.internal_master
------------
Succeeded: 2 (changed=1)
Failed:    2
------------
Total states run:     4
Total run time:  22.798 s
```

### Steps to Reproduce Issue
I think it will be hard to reproduce but here is my state:
https://gist.github.com/Oloremo/c67d9fd877c55ff3b362936d09b1f050

I think it's somehow related to the `formula.java.extract_java` state which uses `archive.extract` to download and extract java from our storage. This operation could take some time, especially with big concurrency.

Please guide me how to provide needed details for this issue.

### Versions Report
```
Salt Version:
           Salt: 2018.3.3

Dependency Versions:
           cffi: 1.6.0
       cherrypy: Not Installed
       dateutil: 2.7.5
      docker-py: 1.10.6
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.7.2
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.5.6
   mysql-python: 1.2.5
      pycparser: 2.14
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 2.7.5 (default, Jul 13 2018, 13:06:57)
   python-gnupg: Not Installed
         PyYAML: 3.10
          PyZMQ: 17.1.2
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 5.1.1
            ZMQ: 4.2.5

System Versions:
           dist: centos 7.5.1804 Core
         locale: UTF-8
        machine: x86_64
        release: 3.10.0-862.3.2.el7.x86_64
         system: Linux
        version: CentOS Linux 7.5.1804 Core
```

**Comment by Ch3LL** [comment]
Can you include the following information:

1. What is the version of your minion? is it also 2018.3.3 similar to master?

2. Can you include a sanitized version of the debug logs on your master and one of the minions when this occurs?

as a workaround for now i bet if you increase the timeout it will return.

**Comment by Oloremo** [comment]
@Ch3LL 
1) Yes it's the same - 2018.3.3
2) Not sure if they're sanitized enough - looks like salt did 3 retries to and when gave up and each retry produce a lot of logs.
Master log: https://gist.github.com/Oloremo/b23eb8e4c15068935998f748250c0207
Minion log: https://gist.github.com/Oloremo/e65e8d3be37c8765f8947f58f09f7ad5

Please note that 3 type of log line was removed in order to shrink logs a bit since they are expected and probably not related:

```
sed -i '/virt-what/d' master.log
sed -i '/Ignoring pillar stack template/d' master.log
sed -i '/Attempting to match/d' master.log
```

> as a workaround for now i bet if you increase the timeout it will return.

Which timeout is it? `gather_job_timeout` or `timeout`?

**Comment by Oloremo** [comment]
Additionally, I tried to run it on fluorine branch with the same result.

**Comment by Ch3LL** [comment]
and if you run `salt '*' saltutil.sync_all` against your 30 minions outside of the orchestration do you see the issue as well?

**Comment by Oloremo** [comment]
@Ch3LL 
After a log debug session I fuguered out the following:
1) I still not understand Salt architecture
2) The root cause of my problem was the salt mine inside pillars. I though it was a good way for dynamyc minions group discavery based on grains. I had 3 mines and 31 minions, during the pillars refresh this produced the 31*3 salt runners at master which gave me a CPU spike. I removed the mine and probles was gone.

This CPU spike probably was a root of the timeouts problems but I still wonder if it's ok. I have m4.xlarge AWS instance for Master, with 4 vcpu. During this spike I had ~200% of cpus util and up to 30-40% of cpu for top salt processes. So it doen't looks like a real cpu bottleneck since I still had some CPU resourses left. 

And If I understand the error correclty it's the minion who was unable to send results to Master, so I wonder what tune could help me in such situation? Like a number of retries or delay between retries or timeout for single try.

**Comment by Ch3LL** [comment]
Can you elaborate on "The root cause of my problem was the salt mine inside pillars." You should be able to define youre mine settings in pillar without a problem. If you can give me a test case that will produce "the `31*3` salt runners at master which gave me a CPU spike" that would help to understand a bit more. does your setup force the salt-master to run `31*3` runner modules?

**Comment by Oloremo** [comment]
@Ch3LL 
I'm sorry, I meant I had a `mine.get` runner(jinja call) inside the Pillars to populate the Pillars based on the mine.get data. Something like this:
```
{% for id, ipaddr in __salt__.saltutil.runner('mine.get', tgt='roles:hadoop-namenode', fun='grains.get', tgt_type='grain') | dictsort() %}
```
Which is obviously a very bad idea.

**Comment by Ch3LL** [comment]
yeah that doesn't surprise me that that doesn't scale very well. It looks like you are trying to get custom data. could you possibly just set and get that from grain data instead?

**Comment by Oloremo** [comment]
@Ch3LL 
That's what I did. But i'd really like to be able to run something like salt mine in Pillars **once** but looks like it's not the way the Pillars suppose to work.

Anyway, looks like this issue were due to salt mine in Pillrs which produce a CPU spike. CPU skipe wasn't really significant in my opinion but looks like that was enought for salt master to timeout.

I feel we could close this issue, unless you want tyo clarify something else.
